src/flag_gems/ops/index_put.py

```diff
@@ -0,0 +1,263 @@
+        args += [
+            f"values_stride{i}," for i in range(index_rank + inp_rank - indices_len)
+        ]
+        args += [
+            "M,",
+            "N,",
+            "IS_ACCUMULATE: tl.constexpr,",
+            "BLOCK_SIZE0: tl.constexpr,",
+            "BLOCK_SIZE1: tl.constexpr,",
+        ]
+        code.writelines(args)
+    code.writeline("):")
+
+    with code.indent():
+        code.writeline("pid0 = tl.program_id(axis=0)")
+        code.writeline("pid1 = tl.program_id(axis=1)")
+        code.writeline(
+            "offset0 = pid0 * BLOCK_SIZE0 + tl.arange(0, BLOCK_SIZE0)[:, None]"
+        )
+        if inp_rank == indices_len:
+            code.writeline("offset1 = pid1 * 1 + tl.arange(0, 1)[None, :]")
+        else:
+            code.writeline(
+                "offset1 = pid1 * BLOCK_SIZE1 + tl.arange(0, BLOCK_SIZE1)[None, :]"
+            )
+        code.newline()
+        code.writeline("cur_idx = offset0")
+        for i in range(index_rank - 1, -1, -1):
+            code.writeline(f"indices_idx{i} = cur_idx % indices0_shape{i}")
+            code.writeline(f"cur_idx = cur_idx // indices0_shape{i}")
+        code.newline()
+        code.writeline("cur_idx = offset1")
+        for i in range(inp_rank - 1, indices_len - 1, -1):
+            code.writeline(f"input_idx{i} = cur_idx % input_shape{i}")
+            code.writeline(f"cur_idx = cur_idx // input_shape{i}")
+        code.newline()
+        code.writeline("mask0 = offset0 < M")
+        for i in range(indices_len):
+            comp = [f"indices_idx{j} * indices{i}_stride{j}" for j in range(index_rank)]
+            code.writeline(
+                f"cur_index{i} = tl.load(indices{i}_ptr + {' + '.join(comp)}, mask=mask0, other=0)"
+            )
+        code.newline()
+        index_mask = [
+            f"(cur_index{i} >= 0) & (cur_index{i} < input_shape{i})"
+            for i in range(indices_len)
+        ]
+        code.writeline(f"index_mask = {' & '.join(index_mask)}")
+        code.writeline("mask1 = offset1 < N")
+        code.writeline("mask = index_mask & mask0 & mask1")
+        code.newline()
+        comp = [f"cur_index{i} * input_stride{i}" for i in range(indices_len)]
+        comp += [
+            f"input_idx{i} * input_stride{i}" for i in range(indices_len, inp_rank)
+        ]
+        code.writeline(f"input_offset = {' + '.join(comp)}")
+        comp = [f"indices_idx{i} * values_stride{i}" for i in range(index_rank)]
+        comp += [
+            f"input_idx{indices_len+i} * values_stride{index_rank+i}"
+            for i in range(inp_rank - indices_len)
+        ]
+        code.writeline(f"values_offset = {' + '.join(comp)}")
+        code.newline()
+        code.writeline("cur_value = tl.load(values_ptr + values_offset, mask=mask)")
+        code.writeline("if IS_ACCUMULATE:")
+        with code.indent():
+            code.writeline(
+                "tl.atomic_add(input_ptr + input_offset, cur_value, mask=mask)"
```

Comment:
Since there is no thread conflict(am i right?), I’m curious which approach is faster: using `tl.atomic_add`, or explicitly loading dst, performing the accumulation manually, and then storing the result with `tl.store`?

Reply:
Because there may be duplicate indexes, so need to use `tl.atomic_add`